packages/rum-recorder/src/domain/rrweb-snapshot/snapshot.ts

```diff
@@ -0,0 +1,589 @@
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable @typescript-eslint/prefer-regexp-exec */
+import {
+  SerializedNode,
+  SerializedNodeWithId,
+  NodeType,
+  Attributes,
+  INode,
+  IdNodeMap,
+  MaskInputOptions,
+  SlimDOMOptions,
+} from './types'
+
+const tagNameRegex = RegExp('[^a-z1-6-_]')
+
+export const IGNORED_NODE = -2
+
+let nextId = 1
+function genId(): number {
+  return nextId++
+}
+
+export function cleanupSnapshot() {
+  // allow a new recording to start numbering nodes from scratch
+  nextId = 1
+}
+
+function getValidTagName(tagName: string): string {
+  const processedTagName = tagName.toLowerCase().trim()
+
+  if (tagNameRegex.test(processedTagName)) {
+    // if the tag name is odd and we cannot extract
+    // anything from the string, then we return a
+    // generic div
+    return 'div'
+  }
+
+  return processedTagName
+}
+
+function getCssRulesString(s: CSSStyleSheet): string | null {
+  try {
+    const rules = s.rules || s.cssRules
+    return rules ? Array.from(rules).map(getCssRuleString).join('') : null
+  } catch (error) {
+    return null
+  }
+}
+
+function getCssRuleString(rule: CSSRule): string {
+  return isCSSImportRule(rule) ? getCssRulesString(rule.styleSheet) || '' : rule.cssText
+}
+
+function isCSSImportRule(rule: CSSRule): rule is CSSImportRule {
+  return 'styleSheet' in rule
+}
+
+function extractOrigin(url: string): string {
```

Comment:
could we replace it by [getOrigin(url: string)](https://github.com/DataDog/browser-sdk/blob/master/packages/core/src/tools/urlPolyfill.ts#L19)?

Reply:
`getOrigin` seem more complicated (builds an `a` element with `buildUrl`) compared to this implementation.
I'm not comfortable enough with JS/TS to guarantee the behavior is the same, so I'd rather not touch the code atm 🤔

Reply:
sure, it could worth looking into that in a separate task though

Reply:
IMO for now it would make sense to keep the current implementation. As @vlad-mh stated, it is simpler and may be faster than `getOrigin`. Using it to transform big stylesheets seems a bit perf critical. Agreed to look into it in a future task.